### PR TITLE
Refactor Consensus

### PR DESCRIPTION
**Pull request by nbougalis**
Classes implementing the consensus process on Ripple are cleaned
up in preparation for modularizations and compartmentalization.

Functions and state related to inter-round consensus are moved out
of NetworkOPs and into Consensus, where they are more effectively
isolated.

Some member functions are changed to free functions and some free
functions have their scope reduced to specific translation units.

* Track inter-round consensus state using new Consensus object
* Devirtualize interfaces
* Reduce NetworkOPs, Consensus and LedgerConsensus interfaces
* Add comments

The first commit was previously reviewed, but feel free to eyeball it. Reviewers: @vinniefalco, @JoelKatz, @HowardHinnant, @ximinez

### CODE REVIEW

**Comment by ximinez** [comment]
Travis failure is valid. Build error on nounity build. https://travis-ci.org/ripple/rippled/jobs/68054407

**Comment by nbougalis** [comment]
Oh nounity build... you vex me so! Will fix.

**Comment by ximinez** [comment]
Looks good, other than the broken build, obviously.

**Comment by ximinez** [comment]
:+1:

**Comment by HowardHinnant** [comment]
:+1: 
I'm in the middle of removing `Serializer` from `SHAMapItem`.  There are some serious merge issues between that work and this.  I'm going to table my work until this gets into develop.

File: src/ripple/app/ledger/Consensus.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+public:
+    using Proposals = hash_map <NodeID, std::deque<LedgerProposal::pointer>>;
+
+    virtual
+    ~Consensus () = default;
+
+    /** Returns whether we are issuing proposals currently. */
+    virtual
+    bool
+    isProposing () const = 0;
```
Comment (vinniefalco) [inline src/ripple/app/ledger/Consensus.h]: Still abstract?

File: src/ripple/app/ledger/impl/ConsensusImp.h [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+#include <ripple/app/ledger/LedgerConsensus.h>
+#include <ripple/app/misc/FeeVote.h>
+#include <ripple/protocol/STValidation.h>
+#include <ripple/shamap/SHAMap.h>
+#include <beast/utility/Journal.h>
+
+namespace ripple {
+
+/** Implements the consensus process and provides inter-round state. */
+class ConsensusImp
```
Comment (vinniefalco) [inline src/ripple/app/ledger/impl/ConsensusImp.h]: What's the difference between a `Consensus` and a `LedgerConsensus`? They sound quite similar.

File: src/ripple/app/ledger/LedgerConsensus.h [hunk 3]
```diff
@@ -41,7 +41,7 @@ namespace ripple {
 class LedgerConsensus
 {
 public:
-    virtual ~LedgerConsensus() = 0;
+    virtual ~LedgerConsensus() = default;
 
     virtual Json::Value getJson (bool full) = 0;
```
Comment (vinniefalco) [inline src/ripple/app/ledger/LedgerConsensus.h]: Still abstract?

File: src/ripple/app/tx/LocalTxs.h [hunk 4]
```diff
@@ -32,16 +33,13 @@ namespace ripple {
...
 
-    static std::unique_ptr<LocalTxs> New ();
-
     // Add a new local transaction
     virtual void push_back (LedgerIndex index, STTx::ref txn) = 0;
 
-    // Apply local transactions to a new open ledger
-    virtual void apply (TransactionEngine&) = 0;
+    // Return the set of local transactions to a new open ledger
+    virtual CanonicalTXSet getTxSet () = 0;
```
Comment (vinniefalco) [inline src/ripple/app/tx/LocalTxs.h]: Still abstract?

**Comment by vinniefalco** [comment]
I like that NetworkOPs is smaller. This would be even better if the interfaces were de-virtualized but that can wait for another pull request. I'm assuming that the logic for the large blocks of moved code was not changed. Given that,
:+1:

**Comment by JoelKatz** [comment]
:+1:

**Pull request closed without merging**